pkg/staking/stakingcontract/contract_test.go

```diff
@@ -0,0 +1,278 @@
+		if err != nil {
+			t.Fatal(err)
+		}
+		stakedAmount, err := contract.GetStake(ctx, addr)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if stakedAmount.Cmp(big.NewInt(13)) == 0 {
+			t.Fatalf("expected %v, got %v", big.NewInt(3), stakedAmount)
+		}
+	})
+
+	t.Run("insufficient stake amount", func(t *testing.T) {
+		t.Parallel()
+
+		totalAmount := big.NewInt(102400)
+		prevStake := big.NewInt(0)
+
+		contract := New(owner, stakingAddress, bzzTokenAddress,
+			transactionMock.New(
+				transactionMock.WithCallFunc(func(ctx context.Context, request *transaction.TxRequest) (result []byte, err error) {
+					if *request.To == bzzTokenAddress {
+						return totalAmount.FillBytes(make([]byte, 32)), nil
+					}
+					if *request.To == stakingAddress {
+						return prevStake.FillBytes(make([]byte, 32)), nil
+					}
+					return nil, errors.New("unexpected call")
+				})),
+			nonce)
+
+		err := contract.DepositStake(ctx, big.NewInt(0), addr)
+		if !errors.Is(err, ErrInsufficientStakeAmount) {
+			t.Fatal(fmt.Errorf("wanted %w, got %v", ErrInsufficientStakeAmount, err))
+		}
+	})
+
+	t.Run("insufficient funds", func(t *testing.T) {
+		t.Parallel()
+
+		totalAmount := big.NewInt(0)
+		prevStake := big.NewInt(0)
+
+		contract := New(owner, stakingAddress, bzzTokenAddress,
+			transactionMock.New(
+				transactionMock.WithCallFunc(func(ctx context.Context, request *transaction.TxRequest) (result []byte, err error) {
+					if *request.To == bzzTokenAddress {
+						return totalAmount.FillBytes(make([]byte, 32)), nil
+					}
+					if *request.To == stakingAddress {
+						return prevStake.FillBytes(make([]byte, 32)), nil
+					}
+					return nil, errors.New("unexpected call")
+				})),
+			nonce)
+
+		err := contract.DepositStake(ctx, big.NewInt(10), addr)
+		if !errors.Is(err, ErrInsufficientFunds) {
+			t.Fatal(fmt.Errorf("wanted %w, got %v", ErrInsufficientFunds, err))
+		}
+	})
+
+	t.Run("invalid call data", func(t *testing.T) {
+		t.Parallel()
+
+		_, err := stakingABI.Pack("invalidMethod", owner, nonce, stakedAmount)
+		if err == nil {
+			t.Fatal("expected 'method 'invalidMethod' not found'")
+		}
+	})
+
+	t.Run("insufficient funds", func(t *testing.T) {
+		t.Parallel()
+
+		totalAmount := big.NewInt(0)
+		prevStake := big.NewInt(0)
+
+		contract := New(owner, stakingAddress, bzzTokenAddress,
+			transactionMock.New(
+				transactionMock.WithCallFunc(func(ctx context.Context, request *transaction.TxRequest) (result []byte, err error) {
+					if *request.To == bzzTokenAddress {
+						return totalAmount.FillBytes(make([]byte, 32)), nil
+					}
+					if *request.To == stakingAddress {
+						return prevStake.FillBytes(make([]byte, 32)), nil
+					}
+					return nil, errors.New("unexpected call")
+				})),
+			nonce)
+
+		err := contract.DepositStake(ctx, big.NewInt(10), addr)
+		if !errors.Is(err, ErrInsufficientFunds) {
+			t.Fatal(fmt.Errorf("wanted %w, got %v", ErrInsufficientStakeAmount, err))
+		}
+	})
```

Comment:
There are still missing tests I think. We need to fail the WithSendFunc, WithWaitForReceiptFunc and make sure all error handling is correct.

We can also fail the CallFunc with some incorrect values and check if the Unpacking error handling is correct.